src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Interfaces/IUserInviteRepository.cs

```diff
@@ -0,0 +1,15 @@
+﻿#nullable enable
+using System;
+using System.Threading.Tasks;
+using GovUk.Education.ExploreEducationStatistics.Admin.Areas.Identity.Data.Models;
+using GovUk.Education.ExploreEducationStatistics.Admin.Models;
+
+namespace GovUk.Education.ExploreEducationStatistics.Admin.Services.Interfaces
+{
+    public interface IUserInviteRepository
+    {
+        public Task<UserInvite> CreateIfNoOtherUserInvite(string email, Role role, Guid createdById);
+
+        public Task<UserInvite> CreateIfNoOtherUserInvite(string email, string roleId, Guid createdById);
```

Comment:
I didn't realise that email was the primary key - I've made the change. 

Generally, I think I'd prefer for methods to be more verbose and accurate in these kinds of instance - the extra words seems like a very small cost to me compared to the added potential for bugs by being concise. If it were up to me, even in this instance I think I'd leave it - I don't see much benefit to being concise beyond aesthetics - and I'll happily trade less bugs for bad aesthetics.

Be interested in what the other devs think about this.

Reply:
You're right, there's not much in it in this instance so I'm not super bothered either way. I've not mentioned it in other places where it's probably better to be more explicit :wink:

Think it's worth elaborating on this a bit further to give you a more concrete rationale for being less explicit. There's a few things I like to think about in terms of naming things:

1. Direct context - does the name need to include things that can already be inferred from the class/module/scope?

 This is fairly simple, avoid repeating what the surrounding context is already telling us. In this example, we're in a `IUserInviteRepository`, so there's no point in a method being called `CreateUserInvite`. This is just redundant to declare as the class/interface name tells us that it interacts with a database (it's a repository), and it's specifically working with `UserInvites`.

2. Method signature - does the name tell us anything that the arguments and return type don't?

A method isn't just it's name, it's also its arguments and return type. A method like `FindById(Guid Id)` is unnecessarily verbose as we only offer one parameter - an ID. It's entirely redundant to include the 'ById' part.

Explicit naming is way more beneficial in dynamic languages where we don't know what the hell we're working with :grimacing: It doesn't mean we need to do in static languages where we have much richer type information.

3. Other methods - does the name conflict with any similar sounding methods in the current scope?

When we have methods that sound or do similar things, we may want to be more explicit. However, if the functions are basically the same, but the arguments are different, we usually just want to fallback on overloading to simplify the API.

When consuming any API, we typically want to avoid burdening the consumer with excessive methods as this creates a burden of choice. It'd be pretty overwhelming to try auto-complete your method and be greeted by tens of methods with slight variations of the same name. 

I usually see overloading as a way to funnel consumers towards using certain methods. If you've ever used any decent library API/SDK, many of them use overloading in this way.

4. General architecture and responsibilities - does the name work with how the class and method is supposed to be used in the bigger picture?

This circles back a bit towards 1, but it's worth just repeating that naming is impacted by the direct context. In our example, we're working with a repository class, so we know that a method like `Create` should create something in the database. However, we should also think about the repository's responsibility and what it _should_ be doing in the grand scheme of things. 

In general, with the service-repository pattern, we want to leave business logic in the service, whilst repositories should be purely concerned with database access. Unfortunately, it can be unclear where one responsibility ends and the other starts (hence why we're even talking about this). 

I'd probably argue that a repository should just plainly allow us to save something into the database without worrying too much about the surrounding rules. From the service's standpoint, it just wants to tell the repository to do something and get some indication of whether that succeeded or not (e.g. from what it returns or throws).

As I mentioned initially, it's a bit of an implementation detail if an email actually gets created in the database or not, especially if it's not creating one because it already exists. It's not important to know how we got to the result, rather that we got to it at all.

Of course, this is something we could probably debate about forever :sweat_smile: 

5. Expressiveness - does the name express the action succinctly and avoid visual clutter? Is it confusing to read?

Whilst being explicit has benefits, it has a clear impact on creating more visual clutter and making things harder to read. For example, if we were being super explicit, we might name our method something silly like `CreateUserInviteIfItDoesNotExistInSQLDatabase`. This is of course ridiculous, but illustrates the point that being more explicit doesn't directly convert into a better experience.

Personally, a good name is short and sweet. It should concisely express an idea, like any good writing. Any additional information in the name needs to have a good reason to exist, otherwise, it's just noise and should be pruned. As Marie Kondo says - if it doesn't spark joy, get rid of it.

Finally, it's worth mentioning that explicitness can just increase confusion. In our example, `CreateIfNotExists` can imply that there is a way to create multiple of the same invite in the database. This isn't the case, so it's a little confusing to imply that something like that is on the table. Whilst not a major deal, it prompts an idea that we don't really need to burden people with.

Anyway, writing this all out ended up being way longer than I anticipated :laughing: I'm probably going to write a blog post about this at some point!